frontend/simple/js/backend/hapi.js

```diff
@@ -135,3 +135,12 @@ export class HapiNamespace extends TrustedNamespace {
 }
 
 export const namespace = new HapiNamespace()
+const api = {
+  '/v1/register': function ({name, value}) {
+    return namespace.register(name, value)
+  },
+  '/v1/lookup': function ({name}) {
+    return namespace.lookup(name)
+  }
+}
+sbp.registerDomain('namespace', api)
```

Comment:
OK, so this comment applies not only to these functions but to all other functions for which this comment applies... heh, if you followed that.

Basically, as we discussed on the phone, the actual implementation of the function should be placed into these API functions, and not delegated elsewhere.

Now, yes, the plan with having a `HapiNamespace` class was to make it easy to switch out from our server-based namespace to a blockchain-based namespace. So the temptation to just call out to the instance here is understandable.

However, as we migrate to a purely SBP-based project, we should take extreme pains to make sure that we get the naming of our selectors and domains right.

Here are some observations, please note how they apply both to this file and all the other files:

- The domain `namespace` is very generic and therefore conveys very little meaning. That is a problem.
- The selectors here don't help with the meaning
- What is missing here is the **kind** of namespace that is being references. **This is a bug!**

With the way this is currently implemented, the implication would be that when we convert to a namecoin namespace the class would be swapped our appropriately and the same API would be used. However, this is a violation of the SBP-contract — that APIs have **one specific functionality that does not change.**

We cannot at one time have an API mean "this registers a name on the server" and at another time mean "this registers a name on a blockchain". For one, the API might be different for the two (as a blockchain might require passing in additional parameters), and for two, it would violate the SBP API guarantee of **doing only one thing for its entire lifespan.**

Therefore, as a remedy, one or more of the following are needed:

- A different name for the domain (I'm thinking `namespace` might be fine for now)
- A different name for the selector path (this is my current recommendation)

**Plus the implementation should be moved into these functions.**

So, that suggests some refactoring and likely getting rid of the `HapiNamespace` and `HapiBackend` backend classes.

For the `namespace` domain, the selectors could be `/v1/hapi-server/register`, `/v1/hapi-server/lookup` etc. In the future we would create another API for each blockchain (e.g. `namespace/v1/ethereum/register` etc.)

For `backend`, they would have similar modifications.

Why else must we do this?

Consider the following scenario:

- A user opens an issue and gives us the output of their log
- The log is by this point entirely based on SBP output, so the selectors that are logged would be (in the unfixed version) `'/v1/register'` etc.

We would have to ask the user for additional info, like the version of the software that they're using, just to figure out whether these log messages represent registrations on the hapi-backend or in some blockchain.

So, please modify this file and all other similar files to "take full advantage of the Selector-based Programming Paradigm", and if you have any questions / comments about this let's do a chat.

Reply:
I totally agree with you. I'm for the different name for the selector.